pkg/server/handler/authutil.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return
+	}
+
+	err = f.DBConn.GetAuth(user.ID.Key, &authInfo)
+
+	return
+}
+
+func (f *UserAuthFetcher) FetchUser(authData skydb.AuthData) (user skydb.Record, err error) {
+	if err = validateAuthData(authData); err != nil {
+		return
+	}
+
+	query := f.buildAuthDataQuery(authData)
+
+	var results *skydb.Rows
+	results, err = f.Database.Query(&query)
+	if err != nil {
+		return
+	}
+	defer results.Close()
+
+	records := []skydb.Record{}
+	for results.Scan() {
+		record := results.Record()
+		records = append(records, record)
+	}
+
+	if err = results.Err(); err != nil {
+		return
+	}
+
+	recordsQueried := len(records)
+	if recordsQueried == 0 {
+		err = skydb.ErrUserNotFound
+		return
+	} else if recordsQueried > 1 {
+		panic(fmt.Errorf("want 1 records queried, got %v", recordsQueried))
+	}
+
+	user = records[0]
+	getAuthDataFromUser(authData, user)
+
+	return
+}
+
+func (f *UserAuthFetcher) buildAuthDataQuery(authData skydb.AuthData) skydb.Query {
+	username, _ := authData["username"].(string)
+	email, _ := authData["email"].(string)
+
+	predicates := []interface{}{}
+
+	makeILikePredicate := func(keyPath string, value string) skydb.Predicate {
+		return skydb.Predicate{
+			Operator: skydb.ILike,
+			Children: []interface{}{
+				skydb.Expression{Type: skydb.KeyPath, Value: keyPath},
+				skydb.Expression{Type: skydb.Literal, Value: value},
+			},
+		}
+	}
+
+	if username != "" {
+		predicates = append(predicates, makeILikePredicate("username", username))
+	}
+
+	if email != "" {
+		predicates = append(predicates, makeILikePredicate("email", email))
+	}
+
+	one := uint64(1)
+	query := skydb.Query{
+		Type: "user",
+		Predicate: skydb.Predicate{
+			Operator: skydb.And,
+			Children: predicates,
+		},
+		Limit: &one,
+	}
+	return query
+}
+
+// createUserWithRecordContext is a context for creating a new user with
+// database record
+type createUserWithRecordContext struct {
+	DBConn       skydb.Conn
+	Database     skydb.Database
+	AssetStore   asset.Store
+	HookRegistry *hook.Registry
+	Context      context.Context
+}
+
+func (ctx *createUserWithRecordContext) execute(info *skydb.AuthInfo, authData skydb.AuthData) skyerr.Error {
+	db := ctx.Database
+	txDB, ok := db.(skydb.Transactional)
+	if !ok {
+		return skyerr.NewError(skyerr.NotSupported, "database impl does not support transaction")
+	}
+
+	txErr := withTransaction(txDB, func() error {
+		// Check if AuthData duplicated only when it is provided
+		// AuthData may be absent, e.g. login with provider
+		if len(authData) != 0 {
+			if err := validateAuthData(authData); err != nil {
+				return err
+			}
+
+			fetcher := newUserAuthFetcher(db, ctx.DBConn)
+			_, _, err := fetcher.FetchAuth(authData)
+			if err == nil {
+				return errUserDuplicated
+			}
+
+			if err != skydb.ErrUserNotFound {
+				return skyerr.MakeError(err)
+			}
+		}
+
+		if err := ctx.DBConn.CreateAuth(info); err != nil {
+			if err == skydb.ErrUserDuplicated {
+				return errUserDuplicated
```

Comment:
Have you considered the case where there is a auth object but the corresponding user record is (for some unknown reason) missing?